letsencrypt/acme/messages2.py

```diff
@@ -0,0 +1,276 @@
+    :param body: Resource body.
+    :type body: Instance of `ResourceBody` (subclass).
+
+    :param str uri: Location of the resource.
+
+    """
+    __slots__ = ('body', 'uri')
+
+
+class ResourceBody(jose.JSONObjectWithFields):
+    """ACME Resource body."""
+
+
+class RegistrationResource(Resource):
+    """Registration resource.
+
+    :ivar body: `Registration`
+    :ivar str uri: URI of the resource.
+    :ivar new_authzr_uri: URI found in the 'next' Link header
+
+    """
+    __slots__ = ('body', 'uri', 'new_authzr_uri', 'terms_of_service')
+
+
+class Registration(ResourceBody):
+    """Registration resource body."""
+
+    # on new-reg key server ignores 'key' and populates it based on
+    # JWS.signature.combined.jwk
+    key = jose.Field('key', omitempty=True, decoder=jose.JWK.from_json)
+    contact = jose.Field('contact', omitempty=True, default=())
+    recovery_token = jose.Field('recoveryToken', omitempty=True)
+    agreement = jose.Field('agreement', omitempty=True)
+
+
+class ChallengeResource(Resource, jose.JSONObjectWithFields):
+    """Challenge resource.
+
+    :ivar body: `.challenges.ChallengeBody`
+    :ivar authzr_uri: URI found in the 'up' Link header.
+
+    """
+    __slots__ = ('body', 'authzr_uri')
+
+    @property
+    def uri(self):  # pylint: disable=missing-docstring,no-self-argument
+        # bug? 'method already defined line None'
+        # pylint: disable=function-redefined
+        return self.body.uri
+
+
+class ChallengeBody(ResourceBody):
+    """Challenge resource body.
+
+    Confusingly, this has a similar name to `.challenges.Challenge`, as
+    well as `.achallanges.AnnotatedChallenge` or
+    `.achallanges.IndexedChallenge`. Use names such as ``challb`` to
+    distinguish instances of this class from ``achall`` or ``ichall``.
+
+    .. todo::
+       This class could be integrated with challenges.Challenge, but
+       this way it would be confusing when compared to acme-spec, where
+       all challenges are presented without 'uri', 'status', or
+       'validated' fields.
+
+    """
+
+    __slots__ = ('chall',)
+    uri = jose.Field('uri')
+    status = jose.Field('status', decoder=Status.from_json)
+    validated = fields.RFC3339Field('validated', omitempty=True)
+
+    def to_json(self):
+        jobj = super(ChallengeBody, self).to_json()
+        jobj.update(self.chall.to_json())
+        return jobj
+
+    @classmethod
+    def fields_from_json(cls, jobj):
+        jobj_fields = super(ChallengeBody, cls).fields_from_json(jobj)
+        jobj_fields['chall'] = challenges.Challenge.from_json(jobj)
+        return jobj_fields
+
+
+class AuthorizationResource(Resource):
+    """Authorization resource.
+
+    :ivar body: `Authorization`
+    :ivar new_cert_uri: URI found in the 'next' Link header
+
+    """
+    __slots__ = ('body', 'uri', 'new_cert_uri')
+
+
+class Authorization(ResourceBody):
+    """Authorization resource body.
+
+    :ivar challenges: `list` of `Challenge`
+
+    """
+
```

Comment:
It might be helpful to add additional documentation regarding the expected types of these instance variables.  When attempting to use this you would have to know that there is a special Status class (when the spec says optional string).  Identifier is self explanatory for someone who knows the code, but the spec says dictionary of string.